service/history/workflow/mutable_state_impl.go

```diff
@@ -6252,3 +6259,259 @@ func (ms *MutableStateImpl) RefreshExpirationTimeoutTask(ctx context.Context) er
 
 	return RefreshTasksForWorkflowStart(ctx, ms, ms.taskGenerator, EmptyVersionedTransition)
 }
+
+func (ms *MutableStateImpl) currentVersionedTransition() *persistencespb.VersionedTransition {
+	if len(ms.executionInfo.TransitionHistory) == 0 {
+		// transition history is not enabled
+		return nil
+	}
+	transitionHistory := ms.executionInfo.TransitionHistory
+	return transitionHistory[len(transitionHistory)-1]
+}
+
+func (ms *MutableStateImpl) ApplyMutation(
+	mutation *persistencespb.WorkflowMutableStateMutation,
+) error {
+	prevExecutionInfoSize := ms.executionInfo.Size()
+	currentVersionedTransition := ms.currentVersionedTransition()
+
+	ms.applySignalRequestedIds(mutation.SignalRequestedIds, mutation.ExecutionInfo)
+	ms.applyTombstones(mutation.SubStateMachineTombstoneBatches, currentVersionedTransition)
+	ms.syncExecutionInfo(ms.executionInfo, mutation.ExecutionInfo)
+
+	// this must be called after syncExecutionInfo because syncExecutionInfo will update executionInfo.UpdateInfos
+	ms.applyUpdatesToUpdateInfos(mutation.UpdatedUpdateInfos, currentVersionedTransition)
+
+	ms.applyUpdatesToSubStateMachines(
+		mutation.UpdatedActivityInfos,
+		mutation.UpdatedTimerInfos,
+		mutation.UpdatedChildExecutionInfos,
+		mutation.UpdatedRequestCancelInfos,
+		mutation.UpdatedSignalInfos,
+		currentVersionedTransition,
+	)
+
+	err := ms.applyUpdatesToStateMachineNodes(mutation.UpdatedSubStateMachines)
+	if err != nil {
+		return err
+	}
+
+	ms.approximateSize += ms.executionInfo.Size() - prevExecutionInfoSize
+	return nil
+}
+
+func (ms *MutableStateImpl) ApplySnapshot(
+	snapshot *persistencespb.WorkflowMutableState,
+) error {
+	prevExecutionInfoSize := ms.executionInfo.Size()
+	currentVersionedTransition := ms.currentVersionedTransition()
+
+	ms.applySignalRequestedIds(snapshot.SignalRequestedIds, snapshot.ExecutionInfo)
+	ms.applyTombstones(snapshot.ExecutionInfo.SubStateMachineTombstoneBatches, currentVersionedTransition)
```

Comment:
This can't work for snapshot. We could be sending a snapshot because we can't track that many tombstones in mutate state. So we have to rely on the full list of activities/timers etc. to figure out which sub state machines needs to be deleted.